src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const updatedState = [];
+  let tempObject = { ...state };
+
+  for (const action of actions) {
+    switch (true) {
+      case action.type === 'addProperties':
+        add(action.extraData);
+        break;
+      case action.type === 'removeProperties':
+        remove(action.keysToRemove);
+        break;
+      case action.type === 'clear':
+        clear();
+        break;
+    }
+  }
+
+  function add(data) {
+    tempObject = { ...tempObject, ...data };
+    updatedState.push({ ...tempObject });
+  }
+
+  function remove(data) {
+    for (const del of data) {
+      delete tempObject[del];
+    }
+    updatedState.push({ ...tempObject });
+  }
+
+  function clear() {
+    tempObject = {};
+    updatedState.push({ ...tempObject });
```

Comment:
The 'clear' function is spreading an empty object which is unnecessary. You can directly push an empty object to 'updatedState' without the spread operation.